### PR TITLE
adding request to support HTTPS_PROXY

### PR DESCRIPTION
**Pull request by refextu**
Not the most elegant way to solve https://github.com/web-push-libs/web-push/issues/280 but adds common used npm request to provide solid request options and HTTPS_PROXY env support.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -33,6 +33,8 @@
     "http_ece": "^0.5.2",
     "jws": "^3.1.3",
     "minimist": "^1.2.0",
+    "request": "^2.81.0",
+    "request-promise-native": "^1.0.4",
```
Comment (gauntface) [inline package.json]: Does request-promise-native not have it's own dependency on request?

Comment (gauntface) [inline package.json]: Also any reason not to use node-fetch for this?

File: src/web-push-lib.js [hunk 2]
```diff
@@ -254,45 +254,32 @@ WebPushLib.prototype.sendNotification =
     }
 
     return new Promise(function(resolve, reject) {
```
Comment (gauntface) [inline src/web-push-lib.js]: Can we remove this out promise since the new API is returning a promise?

**Comment by gauntface** [comment]
Is there any chance of adding a basic test to ensure this works (i.e. set a HTTPS_PROXY environment variable to point to a local server and pass the test if the local server receives a request).

File: package.json (hunk 1 above)
Comment (refextu) [inline package.json]: request-promise-native asks for peer dependency, added to avoid warnings

File: src/web-push-lib.js (hunk 2 above)
Comment (refextu) [inline src/web-push-lib.js]: yes, that should be possible, tried to change as little code as possible

**Comment by marco-c** [comment]
@refextu ping?

**Comment by refextu** [comment]
was in holiday, try to write a test this week, not so easy to setup a portable proxy under test conditions.

**Comment by gauntface** [comment]
@refextu if you get stuck I'm more than happy to help (via chat, email, hangout or PR into your branch).

**Comment by abdulhannanali** [comment]
@gauntface Can I pick this up?

**Comment by gauntface** [comment]
That's up to @refextu

**Comment by marco-c** [comment]
@abdulhannanali I suppose @refextu is not going to work on this anymore, so you can pick this up if you want.

**Comment by Dhappi** [comment]
@marco-c not sure if anyone is working on this issue. If not can i pick it up ?

**Comment by marco-c** [comment]
@Dhappi definitely! I'd love to see this finally completed.

**Comment by Dhappi** [comment]
I tried the branch by @refextu and it seems to work fine with process.env.http_proxy. so do we just need a test? or do we need an options object for defining proxy too ?

**Comment by marco-c** [comment]
> I tried the branch by @refextu and it seems to work fine with process.env.http_proxy. so do we just need a test? or do we need an options object for defining proxy too ?

A test and addressing @gauntface's review comments.

**Comment by marco-c** [comment]
It's a pity that `request` has so many dependencies, it goes against our goal of having a small library.

**Comment by Dhappi** [comment]
i can spend some time looking into an alternate solution to ```request```.

**Comment by gauntface** [comment]
@marco-c is there a reason not to go with `node-fetch`? Generally I'm a big fan of heading in that direction - but I'm also unsure if it has fewer dependencies or not.

**Comment by Dhappi** [comment]
@gauntface i have something working with [http-proxy-agent](https://www.npmjs.com/package/https-proxy-agent) without using request. I like the idea of using node-fetch but i think we would still need a proxy agent for that.

**Comment by gauntface** [comment]
@marco-c shall we close this PR in favor of someone else filing a PR implementing the feature with tests?

**Comment by marco-c** [comment]
Yeah, good idea.

**Comment by Hotcapsicum** [comment]
> @refextu if you get stuck I'm more than happy to help (via chat, email, hangout or PR into your branch).
so , if I have proxy ,then I can use web-push in China? I need use proxy to request fcm

**Comment by marco-c** [comment]
@Hotcapsicum there is an option to use a proxy with the library, just look at the options of the `sendNotification` function in the README: https://github.com/web-push-libs/web-push.

**Pull request closed without merging**